File: examples/c/mnist_slp.c
#include <stdint.h>

#include <crystalnet.h>

// y = softmax(flatten(x) * w + b)
s_model_t *slp(context_t *ctx, const shape_t *image_shape, uint32_t arity)
{
    symbol x = var(ctx, image_shape);
    symbol x_ = reshape(ctx, mk_shape(ctx, 1, shape_dim(image_shape)), x);
    symbol w = covar(ctx, mk_shape(ctx, 2, shape_dim(image_shape), arity));
    symbol b = covar(ctx, mk_shape(ctx, 1, arity));

    symbol op1 = apply(ctx, op_mul, (symbol[]){x_, w});
    symbol op2 = apply(ctx, op_add, (symbol[]){op1, b});
    symbol op3 = apply(ctx, op_softmax, (symbol[]){op2});

    return make_s_model(ctx, x, op3);
}

int main()
{
    const uint32_t width = 28;
    const uint32_t height = 28;
    const uint32_t n = 10;
    context_t *ctx = new_context();
    const shape_t *image_shape = mk_shape(ctx, 2, width, height);
    s_model_t *model = slp(ctx, image_shape, n);
    s_trainer_t *trainer = new_s_trainer(model, op_xentropy, opt_sgd);
    dataset_t *ds1 = load_mnist("train");
    dataset_t *ds2 = load_mnist("t10k");
    s_experiment(trainer, ds1, ds2, 10000);
    del_dataset(ds1);
    del_dataset(ds2);
    del_s_trainer(trainer);
    del_context(ctx);
    return 0;
}

File: examples/python/mnist_slp.py
#!/usr/bin/env python3

# TODO: implenent

import os
import sys
path = os.path.join(os.path.dirname(__file__), '../..', 'langs/python')
sys.path.append(path)
import crystalnet as c

print(c.version())
image_shape = c.Shape(28, 28)
print(image_shape)


def slp(image_shape: c.Shape, arity: int):
    x = c.var(image_shape)
    x_ = c.reshape(x, c.Shape(image_shape.dim()))
    w = c.covar(c.Shape(image_shape.dim(), arity))
    b = c.covar(c.Shape(arity))
    y = c.apply(c.mul, x_, w)
    z = c.apply(c.add, y, b)
    o = c.apply(c.softmax, z)
    return c.Model(x, o)


# m = slp(image_shape, 10)
